cpp/include/cudf/detail/gather.cuh

```diff
@@ -0,0 +1,255 @@
+*
+* A negative index `i` is transformed to `i + size`, where `size` is
+* the number of elements in the column being gathered from.
+* Allowable values for the index `i` are in the range `[-size, size)`.
+* Thus, when gathering from a column of size `10`, the index `-1`
+* is transformed to `9` (i.e., the last element), `-2` is transformed
+* to `8` (the second-to-last element) and so on.
+* Positive indices are unchanged by this transformation.
+*---------------------------------------------------------------------------**/
+template <typename map_type>
+struct index_converter : public thrust::unary_function<map_type,map_type>
+{
+  index_converter(size_type n_rows)
+  : n_rows(n_rows) {}
+
+  __device__
+  map_type operator()(map_type in) const
+  {
+    return ((in % n_rows) + n_rows) % n_rows;
+  }
+  size_type n_rows;
+};
+
+
+template <typename MapIterator>
+std::unique_ptr<table>
+gather(table_view const& source_table, MapIterator gather_map_begin,
+       MapIterator gather_map_end, bool check_bounds = false,
+       bool ignore_out_of_bounds = false,
+       bool allow_negative_indices = false,
+       rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+       cudaStream_t stream = 0)
+{
+  auto source_n_cols = source_table.num_columns();
+  auto source_n_rows = source_table.num_rows();
+  auto num_destination_rows = std::distance(gather_map_begin, gather_map_end);
+
+  std::vector<std::unique_ptr<column>> destination_columns;
+
+  // TODO: Could be beneficial to use streams internally here
+  
+  for (size_type i = 0; i < source_n_cols; i++) {
+    column_view src_col = source_table.column(i);
+    // The data gather for n columns will be put on the first n streams
+    destination_columns.push_back(
+				  cudf::experimental::type_dispatcher(src_col.type(),
+								      column_gatherer{},
+								      src_col,
+								      gather_map_begin,
+								      gather_map_end,
+								      ignore_out_of_bounds,
+								      mr,
+								      stream));
+			
+  }
+
+  std::unique_ptr<table> destination_table = std::make_unique<table>(std::move(destination_columns));
+
+  auto bitmask_kernel =
+    ignore_out_of_bounds ? gather_bitmask_kernel<true, decltype(gather_map_begin)> : gather_bitmask_kernel<false, decltype(gather_map_begin)>;
+
+  int gather_grid_size;
+  int gather_block_size;
+  CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(
+	       &gather_grid_size, &gather_block_size, bitmask_kernel));
+
+
+  auto source_table_view = table_device_view::create(source_table);
+  auto destination_table_view = mutable_table_device_view::create(destination_table->mutable_view());
+
+  bitmask_kernel<<<gather_grid_size, gather_block_size>>>(
+							  *source_table_view,
+							  gather_map_begin,
+							  *destination_table_view);
+
+
+  mutable_table_view dest_view = destination_table->mutable_view();
+
+  // set null_count to UNKNOWN_NULL_COUNT
+  for (size_type i = 0; i < destination_table->num_columns(); i++) {
```

Comment:
Few things:
1. This will already happened here: 
```
  auto destination_table_view = mutable_table_device_view::create(destination_table->mutable_view());
```
and again here:
```
  mutable_table_view dest_view = destination_table->mutable_view();
```
Any time you create a `mutable_column_view` from a `column`, it automatically invalidates the null count. 

2. Unfortunately, you can't actually modify the null count of a `column` from a `mutable_column_view` to that `column`. It's something on my backburner to change in the future. 

3. Didn't the old `bitmask_kernel` also compute the null count for each column? It would be more efficient to do this in the `bitmask_kernel` rather than force kernel launches for `null_count()` on every column in `destination_table`.  If you do this, be sure to use `set_null_count` on the column's themselves, not the `mutable_column_view`s.

Reply:
Is there a way to _explicitly_ set the null count of the columns in a table to `UNKOWN_NULL_COUNT`?

Reply:
You can just do 
```
c.set_null_count(cudf::UNKNOWN_NULL_COUNT);
```

I've debated adding:
```
column::invalidate_null_count();
```
which is just a convenience wrapper around the above.

Reply:
Right, my original issue was that there was no way to grab a `column` from a `table` but we talked about that already and I've added it :)